File: javascript/src/shield.js
/**
 * Shield - Authenticated Symmetric Encryption (Node.js), wire format v4.
 *
 * v4 replaces the previous custom SHA-256 keystream + HMAC construction with a
 * standard AEAD (AES-256-GCM by default, ChaCha20-Poly1305 optional) from Node's
 * built-in `crypto` (OpenSSL). No cryptography is hand-rolled. The wire format
 * matches every other Shield binding byte-for-byte (see tests/v4_test_vectors.json).
 */

const crypto = require('crypto');

// Key derivation iterations (OWASP 2023 floor for PBKDF2-HMAC-SHA256)
const PBKDF2_ITERATIONS = 600000;

// Sizes in bytes
const NONCE_SIZE = 12; // 96-bit AEAD nonce
const TAG_SIZE = 16;   // 128-bit AEAD tag
const SALT_SIZE = 16;

// Version bytes (authenticated, leading byte of the ciphertext)
const VERSION_PASSWORD = 0x03; // 0x03 || suite || salt(16) || nonce(12) || ct||tag
const VERSION_KEY = 0x13;      // 0x13 || suite || nonce(12) || ct||tag

// Cipher-suite identifiers
const SUITE_AES_256_GCM = 0x01;
const SUITE_CHACHA20_POLY1305 = 0x02;

// Inner-plaintext layout: timestamp_ms(8) || pad_len(1) || padding || message
const INNER_HEADER_SIZE = 9;
const MIN_PADDING = 32;
const MAX_PADDING = 128;

// HKDF-Expand info string deriving the AEAD key from the master key
const HKDF_AEAD_INFO = Buffer.from('shield/aead/v4');

/**
 * Derive the AEAD key via HKDF-SHA256-Expand(master, "shield/aead/v4", 32).
 * For a 32-byte output this is a single HKDF block: HMAC-SHA256(master, info || 0x01).
 */
function deriveAeadKey(masterKey) {
    return crypto.createHmac('sha256', masterKey)
        .update(Buffer.concat([HKDF_AEAD_INFO, Buffer.from([0x01])]))
        .digest()
        .slice(0, 32);
}

/** Node cipher algorithm name for a suite byte, or null if unknown. */
function algoForSuite(suite) {
    if (suite === SUITE_AES_256_GCM) return 'aes-256-gcm';
    if (suite === SUITE_CHACHA20_POLY1305) return 'chacha20-poly1305';
    return null;
}

/** AEAD seal: returns ciphertext||tag. */
function aeadSeal(suite, key, nonce, aad, plaintext) {
    const algo = algoForSuite(suite);
    if (algo === null) throw new Error(`unknown cipher suite: ${suite}`);
    const cipher = crypto.createCipheriv(algo, key, nonce, { authTagLength: TAG_SIZE });
    cipher.setAAD(aad, { plaintextLength: plaintext.length });
    const ct = Buffer.concat([cipher.update(plaintext), cipher.final()]);
    return Buffer.concat([ct, cipher.getAuthTag()]);
}

/** AEAD open: returns plaintext, or null on authentication failure / unknown suite. */
function aeadOpen(suite, key, nonce, aad, ctTag) {
    const algo = algoForSuite(suite);
    if (algo === null) return null;
    if (ctTag.length < TAG_SIZE) return null;
    const ct = ctTag.slice(0, ctTag.length - TAG_SIZE);
    const tag = ctTag.slice(ctTag.length - TAG_SIZE);
    try {
        const decipher = crypto.createDecipheriv(algo, key, nonce, { authTagLength: TAG_SIZE });
        decipher.setAAD(aad, { plaintextLength: ct.length });
        decipher.setAuthTag(tag);
        return Buffer.concat([decipher.update(ct), decipher.final()]);
    } catch (e) {
        return null; // authentication failed
    }
}

/** Build AEAD additional data (= wire prefix before the nonce). */
function buildAad(suite, salt) {
    if (salt !== null && salt !== undefined) {
        return Buffer.concat([Buffer.from([VERSION_PASSWORD, suite]), salt]);
    }
    return Buffer.from([VERSION_KEY, suite]);
}

/**
 * Deterministic AEAD seal over fully specified inputs (used for conformance
 * vectors and wrapped by the randomized Shield#encrypt path).
 */
function sealDeterministic(aeadKey, suite, salt, nonce, timestampMs, padLen, padding, plaintext) {
    const aad = buildAad(suite, salt);
    const tsBytes = Buffer.alloc(8);
    tsBytes.writeBigUInt64LE(BigInt(timestampMs));
    const inner = Buffer.concat([tsBytes, Buffer.from([padLen]), padding, plaintext]);
    const ctTag = aeadSeal(suite, aeadKey, nonce, aad, inner);
    return Buffer.concat([aad, nonce, ctTag]);
}

/**
 * Open an AEAD ciphertext, validate the inner layout and freshness window.
 * @returns {Buffer|null}
 */
function openCiphertext(aeadKey, suite, encrypted, aadLen, maxAgeMs) {
    if (encrypted.length < aadLen + NONCE_SIZE + TAG_SIZE) return null;
    const aad = encrypted.slice(0, aadLen);
    const nonce = encrypted.slice(aadLen, aadLen + NONCE_SIZE);
    const ctTag = encrypted.slice(aadLen + NONCE_SIZE);

    const inner = aeadOpen(suite, aeadKey, nonce, aad, ctTag);
    if (inner === null) return null;

    if (inner.length < INNER_HEADER_SIZE) return null;
    const timestampMs = Number(inner.slice(0, 8).readBigUInt64LE());
    const padLen = inner[8];
    if (padLen < MIN_PADDING || padLen > MAX_PADDING) return null;
    const dataStart = INNER_HEADER_SIZE + padLen;
    if (inner.length < dataStart) return null;

    // Freshness window (NOT full replay protection)
    if (maxAgeMs !== null && maxAgeMs !== undefined) {
        const nowMs = Date.now();
        const age = nowMs - timestampMs;
        if (timestampMs > nowMs + 5000 || age > maxAgeMs) return null;
    }

    return inner.slice(dataStart);
}

/** Random padding length in [32, 128] via rejection sampling (no modulo bias). */
function samplePadLen() {
    const padRange = MAX_PADDING - MIN_PADDING + 1; // 97
    for (;;) {
        const val = crypto.randomBytes(1)[0];
        if (val < padRange * Math.floor(256 / padRange)) {
            return (val % padRange) + MIN_PADDING;
        }
    }
}

/**
 * Shield encryption class.
 *
 * @example
 * const shield = new Shield('password', 'github.com');
 * const encrypted = shield.encrypt(Buffer.from('secret'));
 * const decrypted = shield.decrypt(encrypted);
 */
class Shield {
    /**
     * @param {string} password
     * @param {string} service
     * @param {Object} options
     * @param {Buffer} options.salt - Explicit 16-byte salt (default: random per instance)
     * @param {number} options.iterations - PBKDF2 iterations (default 600000)
     * @param {number} options.maxAgeMs - Freshness window ms (default 60000, null = disabled)
     * @param {number} options.suite - Cipher suite (default AES-256-GCM)
     */
    constructor(password, service, options = {}) {
        const salt = options.salt || crypto.randomBytes(SALT_SIZE);
        const iterations = options.iterations || PBKDF2_ITERATIONS;
        const maxAgeMs = options.maxAgeMs !== undefined ? options.maxAgeMs : 60000;
        const suite = options.suite !== undefined ? options.suite : SUITE_AES_256_GCM;

        this._password = Buffer.from(password);
        this._service = Buffer.from(service);
        this._iterations = iterations;
        this._salt = salt;
        this._suite = suite;
        this._keyCache = new Map();

        this._key = this._deriveKey(salt);
        this._aeadKey = deriveAeadKey(this._key);
        this._maxAgeMs = maxAgeMs;
    }

    /** Derive the 32-byte master key for a given salt (cached). */
    _deriveKey(salt) {
        const cacheKey = salt.toString('hex');
        const cached = this._keyCache.get(cacheKey);
        if (cached !== undefined) return cached;
        const key = crypto.pbkdf2Sync(
            this._password,
            Buffer.concat([salt, this._service]),
            this._iterations,
            32,
            'sha256'
        );
        this._keyCache.set(cacheKey, key);
        return key;
    }

    /**
     * Create Shield instance from raw key (no password derivation).
     * @param {Buffer} key - 32-byte symmetric key
     * @param {Object} options - { suite }
     */
    static withKey(key, options = {}) {
        if (key.length !== 32) {
            throw new Error(`Key must be 32 bytes, got ${key.length}`);
        }
        const instance = Object.create(Shield.prototype);
        instance._key = key;
        instance._aeadKey = deriveAeadKey(key);
        instance._suite = options.suite !== undefined ? options.suite : SUITE_AES_256_GCM;
        instance._maxAgeMs = options.maxAgeMs !== undefined ? options.maxAgeMs : 60000;
        // Pre-shared-key mode: no password, no salt.
        instance._password = null;
        instance._service = null;
        instance._iterations = null;
        instance._salt = null;
        instance._keyCache = new Map();
        return instance;
    }

    /**
     * Create Shield with hardware fingerprinting (device-bound encryption).
     * @param {string} password
     * @param {string} service
     * @param {Object} options - { mode }
     */
    static withFingerprint(password, service, options = {}) {
        const { FingerprintMode, collectFingerprint } = require('./fingerprint');
        const mode = options.mode || FingerprintMode.COMBINED;
        const fingerprint = collectFingerprint(mode);
        const combinedPassword = fingerprint ? `${password}:${fingerprint}` : password;
        return new Shield(combinedPassword, service, options);
    }

    /**
     * Encrypt data with a standard AEAD and length obfuscation.
     *
     * Output (password mode): 0x03 || suite || salt(16) || nonce(12) || ct||tag
     * Output (key mode):      0x13 || suite || nonce(12) || ct||tag
     *
     * @param {Buffer} plaintext
     * @returns {Buffer}
     */
    encrypt(plaintext) {
        const nonce = crypto.randomBytes(NONCE_SIZE);
        const padLen = samplePadLen();
        const padding = crypto.randomBytes(padLen);
        return sealDeterministic(
            this._aeadKey, this._suite, this._salt, nonce, Date.now(), padLen, padding, plaintext
        );
    }

    /**
     * Decrypt and verify data, dispatching on the leading authenticated version byte.
     * @param {Buffer} encrypted
     * @returns {Buffer|null}
     */
    decrypt(encrypted) {
        if (encrypted.length < 1) return null;
        const version = encrypted[0];

        if (version === VERSION_PASSWORD) {
            const aadLen = 2 + SALT_SIZE;
            if (encrypted.length < aadLen + NONCE_SIZE + TAG_SIZE) return null;
            if (this._salt === null || this._salt === undefined) return null;
            const suite = encrypted[1];
            const salt = encrypted.slice(2, 2 + SALT_SIZE);
            const master = this._deriveKey(salt);
            const aeadKey = deriveAeadKey(master);
            return openCiphertext(aeadKey, suite, encrypted, aadLen, this._maxAgeMs);
        }
        if (version === VERSION_KEY) {
            const suite = encrypted[1];
            return openCiphertext(this._aeadKey, suite, encrypted, 2, this._maxAgeMs);
        }
        return null;
    }

    /** Get the derived master key (for testing/debugging). */
    get key() {
        return this._key;
    }
}

/**
 * One-shot encrypt with a pre-shared key. Equivalent to Shield.withKey(key).encrypt(data):
 * same authenticated, length-obfuscated v4 wire format and HKDF-derived AEAD key as the
 * instance API and the Rust source of truth.
 */
function quickEncrypt(key, data) {
    return Shield.withKey(key).encrypt(data);
}

/**
 * One-shot decrypt with a pre-shared key. Applies the default 60-second freshness window.
 */
function quickDecrypt(key, encrypted) {
    return Shield.withKey(key).decrypt(encrypted);
}

module.exports = {
    Shield,
    quickEncrypt,
    quickDecrypt,
    // Constants
    PBKDF2_ITERATIONS,
    VERSION_PASSWORD,
    VERSION_KEY,
    SUITE_AES_256_GCM,
    SUITE_CHACHA20_POLY1305,
    SALT_SIZE,
    NONCE_SIZE,
    TAG_SIZE,
    INNER_HEADER_SIZE,
    MIN_PADDING,
    MAX_PADDING,
    // Internal: exposed for interop/conformance tests only, not part of public API
    _deriveAeadKey: deriveAeadKey,
    _sealDeterministic: sealDeterministic,
    _openCiphertext: openCiphertext
};

File: android/shield/src/main/java/ai/dikestra/shield/Shield.kt
package ai.dikestra.shield

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.GeneralSecurityException
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import java.security.MessageDigest

/**
 * Authenticated symmetric encryption for Android (wire format v4).
 *
 * v4 replaces the previous custom SHA-256 keystream + HMAC construction with a
 * standard AEAD (AES-256-GCM by default, ChaCha20-Poly1305 optional) from the JCE.
 * No cryptography is hand-rolled; key derivation uses PBKDF2-HMAC-SHA256 +
 * HKDF-SHA256-Expand. The wire format matches every other Shield binding
 * byte-for-byte (see tests/v4_test_vectors.json).
 *
 * - Password mode: 0x03 || suite(1) || salt(16) || nonce(12) || ciphertext||tag
 * - Key mode:       0x13 || suite(1) || nonce(12) || ciphertext||tag
 *
 * AAD = version || suite || [salt]; inner plaintext =
 * timestamp_ms(8 LE) || pad_len(1) || padding(32-128) || message.
 *
 * Example:
 * ```kotlin
 * val shield = Shield.create("my_password", "github.com")
 * val encrypted = shield.encrypt("secret data".toByteArray())
 * val decrypted = shield.decrypt(encrypted)
 * ```
 */
class Shield private constructor(
    private val key: ByteArray,
    // Password-mode fields (null in pre-shared-key mode).
    private val password: String?,
    private val service: String?,
    private val iterations: Int,
    private val salt: ByteArray?,
    private val maxAgeMs: Long?
) {

    private val aeadKey: ByteArray = deriveAeadKey(key)
    private val suite: Byte = SUITE_AES_GCM
    // Cache of derived master keys keyed by the hex of the 16-byte salt.
    private val keyCache: MutableMap<String, ByteArray> = HashMap()

    init {
        if (salt != null) {
            keyCache[toHex(salt)] = key
        }
    }

    /** Derive the 32-byte master key for a given salt (cached by salt). */
    private fun deriveKey(saltBytes: ByteArray): ByteArray {
        val saltKey = toHex(saltBytes)
        keyCache[saltKey]?.let { return it }
        val pbkdf2Salt = saltBytes + service!!.toByteArray(Charsets.UTF_8)
        val derived = pbkdf2(password!!, pbkdf2Salt, iterations, KEY_SIZE)
        keyCache[saltKey] = derived
        return derived
    }

    companion object {
        private const val DEFAULT_ITERATIONS = 600_000  // OWASP 2023 floor for PBKDF2-HMAC-SHA256
        internal const val NONCE_SIZE = 16  // auxiliary layers; base AEAD uses 12-byte nonce
        internal const val MAC_SIZE = 16
        internal const val KEY_SIZE = 32
        internal const val SALT_SIZE = 16

        // Authenticated version bytes (leading byte of the ciphertext).
        internal const val VERSION_PASSWORD: Byte = 0x03  // 0x03 || suite || salt(16) || nonce(12) || ct||tag
        internal const val VERSION_KEY: Byte = 0x13       // 0x13 || suite || nonce(12) || ct||tag

        // Cipher-suite identifiers.
        internal const val SUITE_AES_GCM: Byte = 0x01
        internal const val SUITE_CHACHA20_POLY1305: Byte = 0x02

        private const val MIN_PADDING = 32
        private const val MAX_PADDING = 128
        private const val DEFAULT_MAX_AGE_MS = 60000L

        // Base-AEAD constants.
        private const val AEAD_NONCE_SIZE = 12
        private const val TAG_SIZE = 16
        private const val INNER_HEADER_SIZE = 9 // timestamp(8) + pad_len(1)
        private val HKDF_AEAD_INFO = "shield/aead/v4".toByteArray(Charsets.UTF_8)

        private val random = SecureRandom()

        /** Create Shield instance from password and service name (password mode). */
        @JvmStatic
        @JvmOverloads
        fun create(
            password: String,
            service: String,
            iterations: Int = DEFAULT_ITERATIONS
        ): Shield {
            val salt = randomBytes(SALT_SIZE)
            val pbkdf2Salt = salt + service.toByteArray(Charsets.UTF_8)
            val key = pbkdf2(password, pbkdf2Salt, iterations, KEY_SIZE)
            return Shield(key, password, service, iterations, salt, DEFAULT_MAX_AGE_MS)
        }

        /** Create Shield with pre-shared key (pre-shared-key mode, no password/salt). */
        @JvmStatic
        fun withKey(key: ByteArray): Shield {
            require(key.size == KEY_SIZE) { "Key must be $KEY_SIZE bytes, got ${key.size}" }
            return Shield(key.copyOf(), null, null, 0, null, DEFAULT_MAX_AGE_MS)
        }

        /** Quick encrypt with pre-shared key (pre-shared-key mode, AES-256-GCM, 0x13). */
        @JvmStatic
        fun quickEncrypt(key: ByteArray, plaintext: ByteArray): ByteArray {
            require(key.size == KEY_SIZE) { "Key must be $KEY_SIZE bytes" }
            return seal(deriveAeadKey(key), SUITE_AES_GCM, null, plaintext)
        }

        /** Quick decrypt with pre-shared key (pre-shared-key mode). */
        @JvmStatic
        fun quickDecrypt(key: ByteArray, ciphertext: ByteArray): ByteArray {
            require(key.size == KEY_SIZE) { "Key must be $KEY_SIZE bytes" }
            if (ciphertext.isEmpty() || ciphertext[0] != VERSION_KEY) {
                throw ShieldException.AuthenticationFailed()
            }
            if (ciphertext.size < 2 + AEAD_NONCE_SIZE + TAG_SIZE) throw ShieldException.CiphertextTooShort()
            return openCiphertext(deriveAeadKey(key), ciphertext[1], ciphertext, 2, null)
        }

        /** AEAD key = HKDF-SHA256-Expand(master, "shield/aead/v4", 32) (single HKDF block). */
        internal fun deriveAeadKey(masterKey: ByteArray): ByteArray =
            hmacSha256(masterKey, HKDF_AEAD_INFO + byteArrayOf(0x01)).copyOf(KEY_SIZE)

        /** Build the AEAD additional data (= wire prefix before the nonce). */
        private fun buildAad(suite: Byte, salt: ByteArray?): ByteArray =
            if (salt != null) byteArrayOf(VERSION_PASSWORD, suite) + salt
            else byteArrayOf(VERSION_KEY, suite)

        private fun samplePadLen(): Int {
            val padRange = MAX_PADDING - MIN_PADDING + 1 // 97
            while (true) {
                val v = random.nextInt() and 0xFF
                if (v < padRange * (256 / padRange)) {
                    return (v % padRange) + MIN_PADDING
                }
            }
        }

        /** Seal with a fresh random nonce, timestamp and padding. */
        private fun seal(aeadKey: ByteArray, suite: Byte, salt: ByteArray?, plaintext: ByteArray): ByteArray {
            val nonce = randomBytes(AEAD_NONCE_SIZE)
            val padLen = samplePadLen()
            val padding = randomBytes(padLen)
            return sealDeterministic(aeadKey, suite, salt, nonce, System.currentTimeMillis(),
                padLen, padding, plaintext)
        }

        /**
         * Deterministic AEAD seal over fully specified inputs (used for conformance
         * vectors and wrapped by the randomized seal).
         */
        internal fun sealDeterministic(
            aeadKey: ByteArray, suite: Byte, salt: ByteArray?, nonce: ByteArray,
            timestampMs: Long, padLen: Int, padding: ByteArray, plaintext: ByteArray
        ): ByteArray {
            val aad = buildAad(suite, salt)
            val tsBytes = ByteArray(8)
            ByteBuffer.wrap(tsBytes).order(ByteOrder.LITTLE_ENDIAN).putLong(timestampMs)
            val inner = tsBytes + byteArrayOf(padLen.toByte()) + padding + plaintext
            val ctTag = aeadSeal(suite, aeadKey, nonce, aad, inner)
            return aad + nonce + ctTag
        }

        /**
         * Open an AEAD ciphertext, validate the inner layout and freshness window.
         * aadLen is the offset of the nonce (= len(version||suite||[salt])).
         */
        internal fun openCiphertext(aeadKey: ByteArray, suite: Byte, encrypted: ByteArray, aadLen: Int, maxAgeMs: Long?): ByteArray {
            if (encrypted.size < aadLen + AEAD_NONCE_SIZE + TAG_SIZE) throw ShieldException.CiphertextTooShort()
            val aad = encrypted.copyOfRange(0, aadLen)
            val nonce = encrypted.copyOfRange(aadLen, aadLen + AEAD_NONCE_SIZE)
            val ctTag = encrypted.copyOfRange(aadLen + AEAD_NONCE_SIZE, encrypted.size)

            val inner = aeadOpen(suite, aeadKey, nonce, aad, ctTag)

            if (inner.size < INNER_HEADER_SIZE) throw ShieldException.AuthenticationFailed()
            val tsBytes = inner.copyOfRange(0, 8)
            val timestampMs = ByteBuffer.wrap(tsBytes).order(ByteOrder.LITTLE_ENDIAN).getLong()
            val padLen = inner[8].toInt() and 0xFF
            if (padLen < MIN_PADDING || padLen > MAX_PADDING) throw ShieldException.AuthenticationFailed()
            val dataStart = INNER_HEADER_SIZE + padLen
            if (inner.size < dataStart) throw ShieldException.CiphertextTooShort()

            if (maxAgeMs != null) {
                val nowMs = System.currentTimeMillis()
                val age = nowMs - timestampMs
                if (timestampMs > nowMs + 5000 || age > maxAgeMs) {
                    throw ShieldException.AuthenticationFailed()
                }
            }

            return inner.copyOfRange(dataStart, inner.size)
        }

        /** AEAD seal: returns ciphertext||tag. */
        private fun aeadSeal(suite: Byte, key: ByteArray, nonce: ByteArray, aad: ByteArray, plaintext: ByteArray): ByteArray {
            return try {
                val cipher = aeadCipher(suite, Cipher.ENCRYPT_MODE, key, nonce)
                cipher.updateAAD(aad)
                cipher.doFinal(plaintext)
            } catch (e: GeneralSecurityException) {
                throw ShieldException.AuthenticationFailed()
            }
        }

        /** AEAD open: returns plaintext, throws on auth failure. */
        private fun aeadOpen(suite: Byte, key: ByteArray, nonce: ByteArray, aad: ByteArray, ctTag: ByteArray): ByteArray {
            return try {
                val cipher = aeadCipher(suite, Cipher.DECRYPT_MODE, key, nonce)
                cipher.updateAAD(aad)
                cipher.doFinal(ctTag)
            } catch (e: GeneralSecurityException) {
                throw ShieldException.AuthenticationFailed()
            }
        }

        private fun aeadCipher(suite: Byte, mode: Int, key: ByteArray, nonce: ByteArray): Cipher {
            return when (suite) {
                SUITE_AES_GCM -> Cipher.getInstance("AES/GCM/NoPadding").apply {
                    init(mode, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_SIZE * 8, nonce))
                }
                SUITE_CHACHA20_POLY1305 -> {
                    // Android runtime registers "ChaCha20/Poly1305/NoPadding"; the JDK
                    // (used by local unit tests) registers "ChaCha20-Poly1305". Try both.
                    val cipher = try {
                        Cipher.getInstance("ChaCha20/Poly1305/NoPadding")
                    } catch (e: GeneralSecurityException) {
                        Cipher.getInstance("ChaCha20-Poly1305")
                    }
                    cipher.init(mode, SecretKeySpec(key, "ChaCha20"), IvParameterSpec(nonce))
                    cipher
                }
                else -> throw GeneralSecurityException("Unknown cipher suite")
            }
        }

        private fun toHex(data: ByteArray): String {
            val sb = StringBuilder(data.size * 2)
            for (b in data) {
                sb.append("0123456789abcdef"[(b.toInt() shr 4) and 0xF])
                sb.append("0123456789abcdef"[b.toInt() and 0xF])
            }
            return sb.toString()
        }

        private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(key, "HmacSHA256"))
            return mac.doFinal(data)
        }

        private fun randomBytes(length: Int): ByteArray {
            val bytes = ByteArray(length)
            random.nextBytes(bytes)
            return bytes
        }

        private fun pbkdf2(password: String, salt: ByteArray, iterations: Int, keyLength: Int): ByteArray {
            val spec = PBEKeySpec(password.toCharArray(), salt, iterations, keyLength * 8)
            val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
            return factory.generateSecret(spec).encoded
        }
    }

    /** Encrypt plaintext (password or pre-shared-key mode). */
    fun encrypt(plaintext: ByteArray): ByteArray = seal(aeadKey, suite, salt, plaintext)

    /**
     * Decrypt and verify, dispatching on the leading authenticated version byte.
     * Hard-rejects any unknown version byte (no legacy heuristic fallback).
     */
    fun decrypt(encrypted: ByteArray): ByteArray {
        if (encrypted.isEmpty()) throw ShieldException.CiphertextTooShort()

        return when (encrypted[0]) {
            VERSION_PASSWORD -> {
                if (salt == null) {
                    throw ShieldException.AuthenticationFailed()
                }
                val aadLen = 2 + SALT_SIZE
                if (encrypted.size < aadLen + AEAD_NONCE_SIZE + TAG_SIZE) throw ShieldException.CiphertextTooShort()
                val msgSuite = encrypted[1]
                val msgSalt = encrypted.copyOfRange(2, 2 + SALT_SIZE)
                val derivedKey = deriveKey(msgSalt)
                val derivedAead = deriveAeadKey(derivedKey)
                openCiphertext(derivedAead, msgSuite, encrypted, aadLen, maxAgeMs)
            }
            VERSION_KEY -> {
                if (encrypted.size < 2 + AEAD_NONCE_SIZE + TAG_SIZE) throw ShieldException.CiphertextTooShort()
                openCiphertext(aeadKey, encrypted[1], encrypted, 2, maxAgeMs)
            }
            else -> throw ShieldException.AuthenticationFailed()
        }
    }
}

/**
 * Shield cryptographic utility functions.
 */
object ShieldUtils {
    const val KEY_SIZE = 32

    /**
     * Generate cryptographically secure random bytes.
     */
    @JvmStatic
    fun randomBytes(length: Int): ByteArray {
        return ByteArray(length).also { SecureRandom().nextBytes(it) }
    }

    /**
     * SHA256 hash.
     */
    @JvmStatic
    fun sha256(data: ByteArray): ByteArray {
        return MessageDigest.getInstance("SHA-256").digest(data)
    }

    /**
     * Constant-time byte array comparison.
     */
    @JvmStatic
    fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        var result = 0
        for (i in a.indices) {
            result = result or (a[i].toInt() xor b[i].toInt())
        }
        return result == 0
    }

    /**
     * Securely wipe sensitive data from memory.
     */
    @JvmStatic
    fun secureWipe(data: ByteArray) {
        data.fill(0)
    }

    /**
     * PBKDF2-SHA256 key derivation.
     */
    @JvmStatic
    fun pbkdf2(password: String, salt: ByteArray, iterations: Int, keyLength: Int): ByteArray {
        val spec = javax.crypto.spec.PBEKeySpec(
            password.toCharArray(),
            salt,
            iterations,
            keyLength * 8
        )
        val factory = javax.crypto.SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }
}

/**
 * Shield exceptions.
 */
sealed class ShieldException(message: String) : Exception(message) {
    class InvalidKeySize : ShieldException("Invalid key size")
    class CiphertextTooShort : ShieldException("Ciphertext too short")
    class AuthenticationFailed : ShieldException("Authentication failed")
    class ReplayDetected : ShieldException("Replay attack detected")
    class OutOfOrder : ShieldException("Out of order message")
    class TokenExpired : ShieldException("Token expired")
    class InvalidToken : ShieldException("Invalid token")
    class SessionExpired : ShieldException("Session expired")
}
